internals/overlord/logstate/puller.go

```diff
@@ -0,0 +1,174 @@
+)
+
+// logPuller handles pulling logs from a single iterator and sending to the
+// main control loop.
+type logPuller struct {
+	iterator servicelog.Iterator
+	entryCh  chan<- servicelog.Entry
+
+	tomb tomb.Tomb
+}
+
+// loop pulls logs off the iterator and sends them on the entryCh.
+// The loop will terminate:
+//   - if the puller's context is cancelled
+//   - once the ringbuffer is closed and the iterator finishes reading all
+//     remaining logs.
+func (p *logPuller) loop() error {
+	defer func() { _ = p.iterator.Close() }()
+
+	parser := servicelog.NewParser(p.iterator, parserSize)
+	for p.iterator.Next(p.tomb.Dying()) {
+		for parser.Next() {
+			if err := parser.Err(); err != nil {
+				return err
+			}
+
+			select {
+			case p.entryCh <- parser.Entry():
+			case <-p.tomb.Dying():
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+// pullerGroup represents a group of logPullers, and provides methods for a
+// gatherer to manage logPullers (dynamically add/remove, kill all, wait for
+// all to finish).
+type pullerGroup struct {
+	targetName string
+
+	// Currently active logPullers, indexed by service name
+	pullers map[string]*logPuller
+	// Mutex for pullers map
+	mu sync.RWMutex
+}
+
+func newPullerGroup(targetName string) *pullerGroup {
+	pg := &pullerGroup{
+		targetName: targetName,
+		pullers:    map[string]*logPuller{},
+	}
+	return pg
+}
+
+func (pg *pullerGroup) Add(serviceName string, buffer *servicelog.RingBuffer, entryCh chan<- servicelog.Entry) {
+	lp := &logPuller{
+		iterator: buffer.TailIterator(),
+		entryCh:  entryCh,
+	}
+	lp.tomb.Go(func() error {
+		err := lp.loop()
+		// Once the puller exits, we should remove it from the group.
+		//pg.Remove(serviceName) - this seems to cause deadlock though
+		return err
+	})
+
+	// There shouldn't already be a puller for this service, but if there is,
+	// shut it down first and wait for it to die.
+	pg.Remove(serviceName)
+
+	pg.mu.Lock()
+	defer pg.mu.Unlock()
+	pg.pullers[serviceName] = lp
+}
+
+// List returns a list of all service names for which we have a currently
+// active puller.
+func (pg *pullerGroup) List() []string {
+	pg.mu.RLock()
+	defer pg.mu.RUnlock()
+
+	var svcs []string
+	for svc := range pg.pullers {
+		svcs = append(svcs, svc)
+	}
+	return svcs
+}
+
+func (pg *pullerGroup) Remove(serviceName string) {
+	pg.mu.RLock()
+	puller, pullerExists := pg.pullers[serviceName]
+	pg.mu.RUnlock()
+
+	if !pullerExists {
+		return
+	}
+
+	puller.tomb.Kill(nil)
+	err := puller.tomb.Wait()
+	if err != nil {
+		logger.Noticef("Error from log puller: %v", err)
+	}
+
+	pg.mu.Lock()
+	defer pg.mu.Unlock()
+	delete(pg.pullers, serviceName)
+}
+
+func (pg *pullerGroup) KillAll() {
+	pg.mu.RLock()
+	defer pg.mu.RUnlock()
+
+	for _, puller := range pg.pullers {
+		puller.tomb.Kill(nil)
+	}
+}
+
+// Done returns a channel which can be waited on until all pullers have finished.
+func (pg *pullerGroup) Done() <-chan struct{} {
+	pg.mu.RLock()
+	pullers := pg.pullers
```

Comment:
This is read-locking the access to the map field from pg, but then all actual accesses to the map are done concurrently in a goroutine below, which is a race condition that needs addressing.

There's also a second problem here: such Done channels are used cheaply to verify the conclusion of work. If you ook inside tomb (recommended), you'll notice that this literally a single channel throughout the entire lifetime of the object, and alive/dead are simply states of such channels. Here we instead see a goroutine being created everyt ime the method is called, and it blocks every time waiting for everything to get finished. This sort of pattern can easily become logical memory leaks in the system, with goroutines that never get collected. 

Can we please address both the race condition and that second issue at once?